File: medium/34.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    const ouput = [-1, -1];
    let i = 0,
        j = nums.length - 1,
        mid;
    
    while(i < j){
        mid = ~~((i + j) / 2);
        if(nums[mid] < target) i = mid + 1;
        else j = mid;
    }
    if(nums[i] !== target) return ouput;
    ouput[0] = i;
    
    j = nums.length - 1;
    while(i < j){
        mid = ~~((i + j) / 2) + 1;
        if(nums[mid] > target) j = mid - 1;
        else i = mid;
    }
    ouput[1] = j;
    
    return ouput;
}


console.log(searchRange([1, 2, 3, 4, 5], 3.3))
console.log(searchRange([5, 7, 7, 8], 8))
console.log(searchRange([5, 7, 7, 8], 5))
console.log(searchRange([5, 7, 7, 8, 8, 10], 5))
console.log(searchRange([5, 7, 7, 8, 8, 10], 7))
console.log(searchRange([5, 7, 7, 8, 8, 10], 8))
console.log(searchRange([5, 7, 7, 8, 8, 10], 10))
console.log(searchRange([5, 7, 7, 8, 8, 10], 6))
console.log(searchRange([], 0))

File: medium/34.java
/**
    # 34. Find First and Last Position of Element in Sorted Array
 */
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] res = new int[] {-1, -1};
        res[0] = lowerBound(nums, target, 0, nums.length - 1);
        if(res[0] >= nums.length || nums[res[0]] != target) 
            res[0] = -1;
        else
            res[1] = upperBound(nums, target, res[0], nums.length - 1);

        return res;
    }

    private int lowerBound(int[] nums, int target, int l, int r) {
        while(l < r){
            int m = (l + r) / 2;
            if(nums[m] >= target)
                r = m;
            else
                l = m + 1;
        }
        return l;
    }

    private int upperBound(int[] nums, int target, int l, int r) {
        while(l < r){
            int m = (l + r + 1) / 2;
            if(nums[m] <= target)
                l = m;
            else
                r = m - 1;
        }
        return l;
    }
}
